### PR TITLE
feat(sdk): `print`s are displayed in wing console

### LINKED ISSUE #591: Refactor print to use Logger resource
### Summary

Refactor print to use Logger resource

### Feature Spec

The "print" statement in the compiler currently desugars into `console.log` in both preflight and inflight code, but we would like it to use the Logger resource long-term so that it automatically integrates with the Wing console, and so that it can be customized by users.

### Use Cases

_No response_

### Implementation Notes

Some previous discussion here: https://github.com/winglang/wing/issues/50

The high level idea is to transform an AST node for the `print` call to a `cloud.Logger.of(this).print()` call. We will likely need to add support for static methods on resources. It might also be necessary to model `this` in the AST, or to support either https://github.com/winglang/wing/issues/287 or https://github.com/winglang/wing/issues/513.

### Component

Compiler

### LINKED ISSUE #991: The print method should emit traces
As a user I'd like to see logs on Wing Console when I trigger some inflight code which calls the `print` utility function.
Wing Console's log already shows all trace events produced by the WingSDK, so all we are left to do is emit a trace event when `print` is called. 
The discussion over issue #591 is very relevant to this issue. We can't refactor `print` to use the logger resource yet, because `static` is not yet supported in Wing. Therefore we need to take a different approach as suggested in #591 or something else.

### PR DESCRIPTION
**Pull request by eladb**
All applications now have a `Logger` registered to them.

In order to allow inflight code to emit logs, we introduced an *implicit binding* of the application logger's inflight client to the `$logger` symbol during inflight. This means that `this.$logger` always points to the application logger.

Additionally, in order to capture logs that come from 3rd party libraries, we replace `console.log` with `$logger.print`. So basically, any `console.log` call will be included.

NOTE: The `print()` method of `ILogger` is now a synchronous method (all inflight functions are async by default). Not sure exactly what the implications of this are and if it's a safe assumption but for now it seems like it will work.

Refactored the `Function` classes so that all common code (the stuff the produces the handler asset) is now in `FunctionBase` (as well as handling environment variables). This is where I could implement the common behavior that captured `console.log` and mapped it into the logger.

In preflight, we still just emit `console.log` instead of `print`. At a later stage it might make sense to actually call `Logger.of(this).print()` (in *preflight*) instead.

Resolves #991
Resolves #591

*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Comment by eladb** [comment]
Here's the console with some prints... Next step is to distinguish between log levels. Most of the activity in the log should be "trace" level instead of "info"...

https://user-images.githubusercontent.com/598796/210182991-0e8dc083-566e-405d-8706-d2f92cf250ad.mov

**Comment by eladb** [comment]
@Chriscbr let me know what you think about this

**Review by Chriscbr — APPROVED** [review]
Approach looks alright to me! yeah I think hooking into `console.log` is the right high level direction.. I'm not sure the implications of a synchronous inflight method, but maybe it just means we need a "sync" keyword in the language :^)

File: libs/wingc/src/jsify.rs [hunk 1]
```diff
@@ -550,6 +555,10 @@ impl JSifier {
 		let procid = base16ct::lower::encode_string(&Sha256::new().chain_update(&block).finalize());
 		let mut bindings = vec![];
 		let mut capture_names = vec![];
+		
+		// always capture the application logger via $logger
+		capture_names.push("$logger".into());
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingc/src/jsify.rs:560]: yes, updated.

File: libs/wingsdk/test/target-sim/queue.test.ts [hunk 2]
```diff
@@ -58,16 +58,18 @@ test("queue with one subscriber, default batch size of 1", async () => {
...
-    'Sending messages (messages=["A"], subscriber=sim-0).',
-    'Sending messages (messages=["B"], subscriber=sim-0).',
+    'Sending messages (messages=["A"], subscriber=sim-1).',
+    'Sending messages (messages=["B"], subscriber=sim-1).',
     'Invoke (payload="{"messages":["A"]}").',
     'Invoke (payload="{"messages":["B"]}").',
     "wingsdk.cloud.Queue deleted.",
     "wingsdk.cloud.Function deleted.",
+    "wingsdk.cloud.Logger deleted.",
   ]);
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingsdk/test/target-sim/queue.test.ts:73]: I can convert them to standard snapshot tests. That should be fine.

File: libs/wingsdk/src/core/internal.ts [hunk 3]
```diff
@@ -33,22 +40,28 @@ export function makeHandler(
...
-})`
-      );
+
+      const clientMap = Object.entries(clients)
+        .map(([name, client]) => `${name}: ${client.text}`)
+        .join(",\n");
+
+      return NodeJsCode.fromInline(`
+        (function(clients) {
+          console.log = (...args) => clients.$logger.print(...args);
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingsdk/src/core/internal.ts:50]: yes, that's probably a good observation. Let me think about it for a second.

[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/src/core/internal.ts:50]: In some philosophical sense I wonder if the implementation of "logging" is sort of a feature / aspect of the inflight host, and maybe that's where more of this logic should be defined. e.g. assigning `$logger` in here:

https://github.com/winglang/wing/blob/6529da6bd1cd5994a3f16e5df3c9fa99ece20f60/libs/wingsdk/src/target-tf-aws/function.ts#L46-L49

(not sure - open question)

[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingsdk/src/core/internal.ts]: Yes, this is what I eventually did... See https://github.com/winglang/wing/pull/1012/commits/ced5f18f9632855642231b8d2df5b06aea02c33b

**Review by Chriscbr — APPROVED** [review]
🎶

**Pull request merged**